utils/priority_mutex.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2020 Coinbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"sync"
+)
+
+// PriorityMutex is a special type of mutex
+// that allows callers to request priority
+// over other callers. This can be useful
+// if there is a "hot path" in an application
+// that requires lock access.
+//
+// WARNING: It is possible to cause lock starvation
+// if not careful (i.e. only high priority callers
+// ever do work).
+type PriorityMutex struct {
+	high []chan struct{}
+	low  []chan struct{}
+
+	m sync.Mutex
+	l bool
+}
+
+// NewPriorityMutex returns a new *PriorityMutex.
+func NewPriorityMutex() *PriorityMutex {
```

Comment:
Seems to me this constructor is not needed. The zero value of a `PriorityMutex` should serve just fine. (In particular, there's no reason to initialize `high` and `low` to empty slices rather than `nil`.) A caller can just write `pm := new(PriorityMutex)`.

Reply:
Good call! I've never actually used `append` with a `nil` slice: https://play.golang.org/p/MK0keoRtkJJ